src/Reader.cc

```diff
@@ -0,0 +1,185 @@
+    delete this->readerConfig;
+    if (result != pulsar_result_Ok) {
+      SetError(std::string("Failed to create reader: ") + pulsar_result_str(result));
+      return;
+    }
+  }
+  void OnOK() {
+    Napi::Object obj = Reader::constructor.New({});
+    Reader *reader = Reader::Unwrap(obj);
+    reader->SetCReader(this->cReader);
+    this->deferred.Resolve(obj);
+  }
+  void OnError(const Napi::Error &e) { this->deferred.Reject(Napi::Error::New(Env(), e.Message()).Value()); }
+
+ private:
+  Napi::Promise::Deferred deferred;
+  pulsar_client_t *cClient;
+  ReaderConfig *readerConfig;
+  pulsar_reader_t *cReader;
+};
+
+Napi::Value Reader::NewInstance(const Napi::CallbackInfo &info, pulsar_client_t *cClient) {
+  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(info.Env());
+  Napi::Object config = info[0].As<Napi::Object>();
+  ReaderConfig *readerConfig = new ReaderConfig(config);
+  ReaderNewInstanceWorker *wk = new ReaderNewInstanceWorker(deferred, cClient, readerConfig);
+  wk->Queue();
+  return deferred.Promise();
+}
+
+class ReaderReadNextWorker : public Napi::AsyncWorker {
+ public:
+  ReaderReadNextWorker(const Napi::Promise::Deferred &deferred, pulsar_reader_t *cReader,
+                       int64_t timeout = -1)
+      : AsyncWorker(Napi::Function::New(deferred.Promise().Env(), [](const Napi::CallbackInfo &info) {})),
+        deferred(deferred),
+        cReader(cReader),
+        timeout(timeout) {}
+  ~ReaderReadNextWorker() {}
+  void Execute() {
+    pulsar_result result;
+    if (timeout > 0) {
+      result = pulsar_reader_read_next_with_timeout(this->cReader, &(this->cMessage), timeout);
+    } else {
+      result = pulsar_reader_read_next(this->cReader, &(this->cMessage));
+    }
+    if (result != pulsar_result_Ok) {
+      SetError(std::string("Failed to received message ") + pulsar_result_str(result));
+    }
+  }
+  void OnOK() {
+    Napi::Object obj = Message::NewInstance({}, this->cMessage);
+    this->deferred.Resolve(obj);
+  }
+  void OnError(const Napi::Error &e) { this->deferred.Reject(Napi::Error::New(Env(), e.Message()).Value()); }
+
+ private:
+  Napi::Promise::Deferred deferred;
+  pulsar_reader_t *cReader;
+  pulsar_message_t *cMessage;
+  int64_t timeout;
+};
+
+Napi::Value Reader::ReadNext(const Napi::CallbackInfo &info) {
+  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(info.Env());
+  if (info[0].IsUndefined()) {
+    ReaderReadNextWorker *wk = new ReaderReadNextWorker(deferred, this->cReader);
+    wk->Queue();
+  } else {
+    Napi::Number timeout = info[0].As<Napi::Object>().ToNumber();
+    ReaderReadNextWorker *wk = new ReaderReadNextWorker(deferred, this->cReader, timeout.Int64Value());
+    wk->Queue();
+  }
+  return deferred.Promise();
+}
+
+Napi::Value Reader::HasNext(const Napi::CallbackInfo &info) {
+  int value = 0;
+  pulsar_result result = pulsar_reader_has_message_available(this->cReader, &value);
+  if (result != pulsar_result_Ok || value != 1) {
```

Comment:
Shouldn't we throw an exception if the result is not ok?

Reply:
Yes. I think we should throw it.
So, I fixed same as https://github.com/apache/pulsar/blob/master/pulsar-client-go/pulsar/c_reader.go#L163-L169